filestack/src/main/java/com/filestack/android/CloudListAdapter.java

```diff
@@ -10,106 +10,109 @@
 
 class CloudListAdapter extends RecyclerView.Adapter implements
-        SingleObserver<CloudContents>, View.OnClickListener, View.OnLongClickListener {
+        SingleObserver<CloudContents>, View.OnClickListener, View.OnLongClickListener,
+        FsActivity.BackButtonListener {
 
-    private static final double PAGINATION_THRESHOLD = 0.50;
+    private static final double LOAD_TRIGGER = 0.50;
 
-    private final ClientProvider clientProvider;
-    private final String provider;
+    private final FsAndroidClient.Provider clientProvider;
+    private final String sourceId;
 
-    private int layoutId;
     private RecyclerView recyclerView;
+    private int layoutId;
 
-    private ArrayList<CloudItem> items;
-    private ArrayList<Integer> selected;
-    private Integer threshold;
-    private String nextToken;
+    private String currentPath;
+    private boolean isLoading;
     private boolean multiSelectMode;
 
-    CloudListAdapter(ClientProvider clientProvider, String provider) {
+    private ArrayMap<String, ArrayList<CloudItem>> folders;
+    private ArrayMap<String, String> nextTokens;
+    private ArrayList<Integer> selected;
+
+    CloudListAdapter(FsAndroidClient.Provider clientProvider, String sourceId) {
         this.clientProvider = clientProvider;
-        this.provider = provider;
+        this.sourceId = sourceId;
+
+        folders = new ArrayMap<>();
+        nextTokens = new ArrayMap<>();
         selected = new ArrayList<>();
     }
 
     @Override
     public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {
         LayoutInflater inflater = LayoutInflater.from(viewGroup.getContext());
         View listItemView = inflater.inflate(layoutId, viewGroup, false);
-        return new CloudListItemViewHolder(listItemView);
+        return new CloudListViewHolder(listItemView);
     }
 
     @Override
     public void onBindViewHolder(RecyclerView.ViewHolder viewHolder, int i) {
-        CloudListItemViewHolder holder = (CloudListItemViewHolder) viewHolder;
+        ArrayList<CloudItem> items = folders.get(currentPath);
         CloudItem item = items.get(i);
+        CloudListViewHolder holder = (CloudListViewHolder) viewHolder;
 
         holder.setId(i);
         holder.setName(item.getName());
-        String info = String.format(
-                Locale.getDefault(), "%s - %d", item.getMimetype(), item.getSize());
+        String info = String.format("%s - %d", item.getMimetype(), item.getSize());
         holder.setInfo(info);
         holder.setIcon(item.getThumbnail());
         holder.setSelected(selected.contains(i));
         holder.setOnClickListener(this);
         holder.setOnLongClickListener(this);
 
-        if (threshold != null && nextToken != null && i >= threshold ) {
-            threshold = null;
+        String nextToken = nextTokens.get(currentPath);
+        if (!isLoading && nextToken != null && i >= LOAD_TRIGGER * items.size()) {
```

Comment:
This makes me slightly nervous :P I'm sure it's not a problem, but maybe add some parens to delimit the order of ops?

Reply:
@staturecrane Fixed